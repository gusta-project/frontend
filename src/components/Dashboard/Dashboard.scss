.table-bordered {
  .dashboard-highlight {
    border: 2px solid #f00;
  }
}
