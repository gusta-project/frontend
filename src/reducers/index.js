import { combineReducers } from 'redux';

import {
  reducer as application,
  initialState as applicationState
} from './application';
import { reducer as profile, initialState as profileState } from './profile';
import { reducer as toast, initialState as toastState } from './toast';

export const initialState = {
  application: applicationState,
  profile: profileState,
  toast: toastState
};

export default combineReducers({
  application,
  profile,
  toast
});
