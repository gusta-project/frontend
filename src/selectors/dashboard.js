import { createSelector } from 'reselect';

export const getDashboard = (state) => state.dashboard;

export const getDashboardComponent = createSelector(
  getDashboard,
  (dashboard) => dashboard.dashboardComponent
);

export const getMigrations = createSelector(
  getDashboard,
  (dashboard) => dashboard.migrations
);

export const getStats = createSelector(
  getDashboard,
  (dashboard) => dashboard.stats
);
