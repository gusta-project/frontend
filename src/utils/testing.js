import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
