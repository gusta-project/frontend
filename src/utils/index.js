import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
