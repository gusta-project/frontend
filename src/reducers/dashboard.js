import { buildActions } from 'utils';

export const types = buildActions('dashboard', [
  'REQUEST_DASHBOARD',
  'REQUEST_DASHBOARD_SUCCESS',
  'SELECT_DASHBOARD',
  'SELECT_DASHBOARD_SUCCESS',
  'REQUEST_MIGRATIONS',
  'REQUEST_MIGRATIONS_SUCCESS',
  'REQUEST_MIGRATIONS_FAILURE',
  'REQUEST_STATS',
  'REQUEST_STATS_SUCCESS',
  'REQUEST_STATS_FAILURE'
]);

const requestDashboard = () => ({
  type: types.REQUEST_DASHBOARD
});

const requestDashboardSuccess = (dashboardComponent) => ({
  type: types.REQUEST_DASHBOARD_SUCCESS,
  dashboardComponent
});

const selectDashboard = ({ name, item }) => ({
  type: types.SELECT_DASHBOARD,
  name,
  item
});

const selectDashboardSuccess = (dashboardComponent) => ({
  type: types.SELECT_DASHBOARD_SUCCESS,
  dashboardComponent
});

const requestMigrations = () => ({
  type: types.REQUEST_MIGRATIONS
});

const requestMigrationsSuccess = (migrations) => ({
  type: types.REQUEST_MIGRATIONS_SUCCESS,
  migrations
});

const requestMigrationsFailure = (error) => ({
  type: types.REQUEST_MIGRATIONS_FAILURE,
  error
});

const requestStats = () => ({
  type: types.REQUEST_STATS
});

const requestStatsSuccess = (stats) => ({
  type: types.REQUEST_STATS_SUCCESS,
  stats
});

const requestStatsFailure = (error) => ({
  type: types.REQUEST_STATS_FAILURE,
  error
});

export const actions = {
  requestDashboard,
  requestDashboardSuccess,
  selectDashboard,
  selectDashboardSuccess,
  requestMigrations,
  requestMigrationsSuccess,
  requestMigrationsFailure,
  requestStats,
  requestStatsSuccess,
  requestStatsFailure
};

export const initialState = {
  dashboardComponent: {
    name: 'Home',
    item: null
  },
  migrations: [],
  stats: {}
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.SELECT_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.REQUEST_MIGRATIONS_SUCCESS:
      return {
        ...state,
        migrations: action.migrations
      };
    case types.REQUEST_MIGRATIONS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.REQUEST_STATS_SUCCESS:
      return {
        ...state,
        stats: action.stats
      };
    case types.REQUEST_STATS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
