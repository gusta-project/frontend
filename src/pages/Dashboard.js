import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import {
  DashboardMenu as Menu,
  DashboardMain as Main
} from 'components/Dashboard';
import { Container, Row, Col } from 'react-bootstrap';

import { actions as dashboardActions } from 'reducers/dashboard';
import { getDashboardComponent } from 'selectors/dashboard';

export class Dashboard extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired
  };

  componentDidMount() {
    const { actions } = this.props;

    actions.requestStats();
  }

  render() {
    return (
      <Container>
        <Row>
          <Col>
            <h1>Dashboard</h1>
          </Col>
        </Row>
        <Row>
          <Col xs={12} md={3}>
            <Menu />
          </Col>
          <Col xs={12} md={9}>
            <Main />
          </Col>
        </Row>
      </Container>
    );
  }
}

const mapStateToProps = state => ({
  dashboardComponent: getDashboardComponent(state)
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(dashboardActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
