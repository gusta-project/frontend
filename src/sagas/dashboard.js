import { all, call, put, takeLatest } from 'redux-saga/effects';

import request from 'utils/request';
import { actions, types } from 'reducers/dashboard';
import { actions as toastActions } from 'reducers/toast';

function* requestDashboardWorker(dashboardComponent) {
  yield put(actions.requestDashboardSuccess(dashboardComponent));
}

function* selectDashboardWorker({ name, item }) {
  yield put(actions.selectDashboardSuccess({ name, item }));
}

function* requestMigrationsWorker() {
  try {
    const endpoint = {
      url: '/data/version',
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    // update migrations in state or throw an error
    if (result.success) {
      const {
        response: { data }
      } = result;

      yield put(actions.requestMigrationsSuccess(data));
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to get migrations!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestMigrationsFailure(error));
    yield put(
      toastActions.popToast({
        title: 'Error',
        icon: 'times-circle',
        message
      })
    );
  }
}

function* requestStatsWorker() {
  try {
    const endpoint = {
      url: '/stats/dashboard',
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    // update stats in state or throw an error
    if (result.success) {
      const {
        response: { data }
      } = result;

      yield put(actions.requestStatsSuccess(data));
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to get statistics!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestStatsFailure(error));
    yield put(
      toastActions.popToast({
        title: 'Error',
        icon: 'times-circle',
        message
      })
    );
  }
}

function* requestDashboardWatcher() {
  yield takeLatest(types.REQUEST_DASHBOARD, requestDashboardWorker);
}

function* selectDashboardWatcher() {
  yield takeLatest(types.SELECT_DASHBOARD, selectDashboardWorker);
}

function* requestMigrationsWatcher() {
  yield takeLatest(types.REQUEST_MIGRATIONS, requestMigrationsWorker);
}

function* requestStatsWatcher() {
  yield takeLatest(types.REQUEST_STATS, requestStatsWorker);
}

export const workers = {
  requestDashboardWorker,
  selectDashboardWorker,
  requestMigrationsWorker,
  requestStatsWorker
};

export const watchers = {
  requestDashboardWatcher,
  selectDashboardWatcher,
  requestMigrationsWatcher,
  requestStatsWatcher
};

export default function* saga() {
  yield all(Object.values(watchers).map(watcher => watcher()));
}
