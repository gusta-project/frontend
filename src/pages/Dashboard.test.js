import { render } from '@testing-library/react';
import React from 'react';
import configureStore from 'redux-mock-store';

import { initialState } from 'reducers/dashboard';
import Dashboard from './Dashboard';
import { withMemoryRouter, withProvider } from 'utils/testing';

describe('<Dashboard />', () => {
  const mockStore = configureStore();
  const store = mockStore({ dashboard: initialState });
  const RoutedDashboard = withMemoryRouter(Dashboard);
  const ConnectedDashboard = withProvider(RoutedDashboard, store);

  it('renders correctly', () => {
    const { asFragment } = render(<ConnectedDashboard />);

    expect(asFragment()).toMatchSnapshot();
  });
});
