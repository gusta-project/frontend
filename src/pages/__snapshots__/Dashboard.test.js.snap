// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          Dashboard
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-md-3 col-12"
      >
        <div
          class="accordion"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                class="fa tachometer-alt"
              />
                
              <a
                href="#home"
              >
                Dashboard
              </a>
            </div>
          </div>
          <div
            class="card"
          >
            <div
              class="card-header"
            >
              <h3>
                <i
                  class="fa user-shield"
                />
                 Roles
              </h3>
            </div>
            <div
              class="collapse"
            >
              <div
                class="card-body"
              >
                <div>
                  <i
                    class="fa user-shield"
                  />
                    
                  <a
                    href="#roles"
                  >
                    Roles
                  </a>
                </div>
                <div>
                  <i
                    class="fa plus"
                  />
                    
                  <a
                    href="#roles/add"
                  >
                    Add Role
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="card"
          >
            <div
              class="card-header"
            >
              <h3>
                <i
                  class="fa users"
                />
                 Users
              </h3>
            </div>
            <div
              class="collapse"
            >
              <div
                class="card-body"
              >
                <div>
                  <i
                    class="fa users-cog"
                  />
                    
                  <a
                    href="#users"
                  >
                    Users
                  </a>
                </div>
                <div>
                  <i
                    class="fa plus"
                  />
                    
                  <a
                    href="#users/add"
                  >
                    Add User
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="card"
          >
            <div
              class="card-header"
            >
              <h3>
                <i
                  class="fa bezier-curve"
                />
                 Ingredients
              </h3>
            </div>
            <div
              class="collapse"
            >
              <div
                class="card-body"
              >
                <div>
                  <i
                    class="fa eye-dropper"
                  />
                    
                  <a
                    href="#flavors"
                  >
                    Flavors
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="card"
          >
            <div
              class="card-header"
            >
              <h3>
                <i
                  class="fa database"
                />
                 Database
              </h3>
            </div>
            <div
              class="collapse"
            >
              <div
                class="card-body"
              >
                <i
                  class="fa clock"
                />
                  
                <a
                  href="#migrations"
                >
                  Migration History
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-9 col-12"
      >
        <div
          class="card-columns"
        >
          <div
            class="card bg-danger text-white"
          >
            <div
              class="card-header"
            >
              Users: 
            </div>
            <div
              class="card-body"
            >
              <h3>
                Activated Users: 
              </h3>
              <h3>
                Inactive Users: NaN
              </h3>
              <h3>
                Active Tokens: 
              </h3>
            </div>
          </div>
          <div
            class="card bg-primary text-white"
          >
            <div
              class="card-body"
            >
              <h2>
                Tags: 
              </h2>
            </div>
          </div>
          <div
            class="card bg-secondary text-white"
          >
            <div
              class="card-body"
            >
              <h2>
                Vendors: 
              </h2>
            </div>
          </div>
          <div
            class="card bg-info text-white"
          >
            <div
              class="card-header"
            >
              Flavors: 
            </div>
            <div
              class="card-body"
            >
              <h2>
                Tags: 
              </h2>
            </div>
          </div>
          <div
            class="card bg-warning text-white"
          >
            <div
              class="card-header"
            >
              Recipes: 
            </div>
            <div
              class="card-body"
            >
              <h2>
                Tags: 
              </h2>
            </div>
          </div>
        </div>
        <a
          href="#home"
          title="Refresh Stats"
        >
          <i
            class="fa sync"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
